---
output: github_document
---

```{r, echo = FALSE, message = FALSE}
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  collapse = TRUE,
  echo = FALSE,
  comment = "#>",
  fig.path = "man/figures/",
  knitr::opts_chunk$set(dpi=300,fig.width=8)
)
```

# scmetadata
<!-- badges: start -->
[![R build status](https://github.com/rnabioco/scmetadata/workflows/pkgdown/badge.svg)](https://github.com/rnabioco/scmetadata/actions)
[![Last Commit on GitHub](`r paste0("https://img.shields.io/badge/last%20run-", format(format = "%m--%d--%Y", Sys.Date()), "-brightgreen")`)](https://rnabioco.github.io/scmetadata/articles/get_geo.html)
<!-- badges: end -->

Cell-type annotations are frequently excluded from public single cell datasets. This hinders single cell sequencing analysis reproducibility and accessibility. To better describe the issue, we monitor GEO entries monthly (currently set to auto-update at 1AM UTC, 20th of the month), and programmatically determine the fraction of entries with (potentially) usable cell metadata. 

As of the initial presentation of this issue (10--17--2020), the number is a frustratingly low **0.122**.

Please find the latest numbers below, or at [full report page](https://rnabioco.github.io/scmetadata/articles/get_geo.html).

```{r load}
library(tidyverse)
library(GEOquery)
library(here)
date <- format(format = "%m%d%y", Sys.Date())
date_1off <- format(format = "%m%d%y", Sys.Date() - 1)
if (file.exists(here("inst", "extdata", date, paste0("geo_", date, ".rds")))) {
  gds <- readRDS(here("inst", "extdata", date, paste0("geo_", date, ".rds"))) %>% 
    mutate(year = as.numeric(year))
} else {
  gds <- readRDS(here("inst", "extdata", date_1off, paste0("geo_", date_1off, ".rds"))) %>% 
    mutate(year = as.numeric(year))
}
gds_orig <- gds
```

```{r filteragain}
checks <- suppressWarnings(map(gds$geo,
                               function(x) str_detect(str_to_lower(x), 
                                                      "single nuclei|single-nucleus|single cell|single-cell|scrnaseq|scrna-seq|snrnaseq|snrna-seq") %>% any()))
gds <- gds[unlist(checks), ]
```

```{r superseries}
parse_subseries <- function(geolist) {
  tryCatch({rel <- geolist$relation
            rel[str_detect(rel, "^SuperSeries of: ")] %>%
              str_remove("^SuperSeries of: ")},
           error = function(e) {return(NA)})
}

res <- map(gds$geo,
           parse_subseries)
names(res) <- gds$id
res2 <- res[lengths(res) != 0]
res2 <- map(res2,
            function(x) ifelse(x %in% gds$id, x, NA))
res3 <- data.frame(superseries = names(res2))
res3$id <- res2
dict_series <- res3 %>% unnest(id) %>% na.omit()
gds2 <- gds %>% left_join(dict_series) %>% 
  mutate(superseries = ifelse(is.na(superseries),
                              id,
                              superseries)) %>% 
  group_by(superseries) %>% 
  summarize(usable = ifelse(any(usable == "yes"), "yes", "no"))

gds_fixed <- gds %>% inner_join(gds2 %>% select(id = superseries, fixed_usable = usable))
gds <- gds_fixed %>% select(-usable) %>% rename(usable = fixed_usable)
```

```{r fixfalsep}
gds_old <- gds
fp <- read_tsv(here("inst", "extdata", "falsepos.txt"), col_names = F) %>% pull(1)
gds <- gds %>% mutate(usable = ifelse(id %in% fp, "no", usable))
```

Current fraction with metadata: **`r (gds$usable == "yes") %>% mean(na.rm = TRUE)`**.

```{r comp}
gds_comp <- readRDS(here("inst", "extdata", "101520", paste0("geo_", "101520", ".rds"))) %>% 
    mutate(year = as.numeric(year))
id_comp <- gds_comp %>% filter(usable == "no") %>% pull(id)
id_gds <- gds_comp %>% filter(usable == "yes") %>% pull(id)
```

Number of depositions with updated metadata records since description of the issue here: **`r length(intersect(id_gds, id_comp))`**.

```{r frac}
frac <- gds %>% na.omit() %>% 
  group_by(year, usable) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  pivot_wider(names_from = usable, values_from = n) %>% 
  replace_na(list(no = 0, yes = 0)) %>% 
  mutate(fraction = yes/(yes + no), total = yes + no) %>% 
  mutate(fraction = round(fraction, digits = 3))

g <- ggplot(gds %>% na.omit() %>% filter(year >= 2012),
       aes(x = year, fill = usable)) +
  geom_bar() +
  scale_fill_brewer() +
  #scale_fill_viridis_d(option = "plasma") +
  geom_text(data = frac %>% filter(year >= 2012), color = "black", aes(x = year, y = total + max(frac$total) * 0.03, label = fraction, fill = NA), size = 3) +
  cowplot::theme_cowplot() +
  scale_x_continuous(breaks = c(2012:2020)) +
  ylab("total studies") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 14), 
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14))
g + ggtitle("scRNAseq datasets by year")

last12 <- gds %>% na.omit() %>%
  select(year, month) %>% 
  distinct() %>% 
  mutate(month = factor(month, levels = c("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"))) %>% 
  arrange(desc(year), desc(as.numeric(month))) %>% 
  dplyr::slice(1:12) %>%
  mutate(full = str_c(year, month, sep = "-")) %>% 
  pull(full)

last12_df <- gds %>% na.omit() %>%
  mutate(full = str_c(year, month, sep = "-")) %>% 
  filter(full %in% last12) %>% 
  mutate(month = factor(month, levels = c("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"))) %>% 
  arrange(year, as.numeric(month))

last12_frac <- last12_df %>% na.omit() %>% 
  group_by(year, month, usable) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  pivot_wider(names_from = usable, values_from = n) %>% 
  replace_na(list(no = 0, yes = 0)) %>% 
  mutate(fraction = yes/(yes + no), total = yes + no) %>% 
  mutate(fraction = round(fraction, digits = 3)) %>%
  mutate(full = str_c(year, month, sep = "-"))
  
g2 <- ggplot(last12_df %>% mutate(full = factor(full, levels = unique(last12_df$full))),
             aes(x = full, fill = usable)) +
  geom_bar() +
  geom_text(data = last12_frac %>% mutate(full = factor(full, levels = unique(last12_df$full))),
            color = "black", 
            aes(x = full, y = total + max(last12_frac$total) * 0.03, label = fraction, fill = NA), size = 3) +
  scale_fill_brewer() +
  cowplot::theme_cowplot() +
  ylab("total studies") +
  xlab("last 12 months") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 14), 
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14))
g2 + ggtitle("scRNAseq datasets by month")
```
