---
title: 'single cell GEO records'
date: '`r Sys.Date()`'
# package: someta
author:
  - name: Rui Fu
    affiliation: RNA Bioscience Initative, University of Colorado School of Medicine
output:
  BiocStyle::html_document:
    toc_float: true

vignette: >
    %\VignetteIndexEntry{metadata_report}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}  
---

```{r "knitr options", echo = FALSE, message = FALSE, warning = FALSE}
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  echo = FALSE,
  collapse = TRUE,
  fig.align = "center",
  comment = "#>",
  crop = NULL
)
```

```{r load}
library(here)
library(tidyverse)
date <- format(format = "%m%d%y", Sys.Date())
```

# Current GEO query used: 

```
"expression profiling by high throughput sequencing" AND
("single nuclei" OR "single cell" OR "scRNAseq" OR "scRNA-seq" OR "snRNAseq" OR "snRNA-seq")
```

```{r eutil, eval = T}
library(httr)
print(date)
esearch <- "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=gds&term=%22expression%20profiling%20by%20high%20throughput%20sequencing%22%20AND%20(%22single%20nuclei%22%20OR%20%22single%20cell%22%20OR%20%22scRNAseq%22%20OR%20%22scRNA-seq%22%20OR%20%22snRNAseq%22%20OR%20%22snRNA-seq%22)&usehistory=y&retmode=json"
get1 <- GET(esearch)
webenv <- content(get1)$esearchresult$webenv
efetch <- paste0("https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=gds&query_key=1&WebEnv=",
                webenv)
dir.create(here("inst", "extdata", date), showWarnings = FALSE)
GET(
  efetch,
  write_disk(here("inst", "extdata", date, paste0("gds_result_", date, ".txt")))
)
```

# Parse GEO query

```{r entries}
source(here("utils", "geo_query.R"))
gds <- readRDS(here("inst", "extdata", date, paste0("geo_", date, ".rds"))) %>% 
  mutate(year = as.numeric(year))
gds_backup <- readRDS(here("inst", "extdata", "101520", paste0("geo_", "101520", ".rds"))) %>% 
  mutate(year = as.numeric(year))
gds_orig <- gds
```

```{r filteragain}
checks <- suppressWarnings(map(gds$geo,
                               function(x) str_detect(str_to_lower(x), 
                                                      "single nuclei|single-nucleus|single cell|single-cell|scrnaseq|scrna-seq|snrnaseq|snrna-seq") %>% any()))
gds <- gds[unlist(checks), ]
```

Total entries: `r nrow(gds_orig)`

Number of entries filtered out because key words were not found: `r sum(!unlist(checks), na.rm = TRUE)`

```{r fixseries}
parse_subseries <- function(geolist) {
  tryCatch({rel <- geolist$relation
            rel[str_detect(rel, "^SuperSeries of: ")] %>%
              str_remove("^SuperSeries of: ")},
           error = function(e) {return(NA)})
}

res <- map(gds$geo,
           parse_subseries)
names(res) <- gds$id
res2 <- res[lengths(res) != 0]
res2 <- map(res2,
            function(x) ifelse(x %in% gds$id, x, NA))
res3 <- data.frame(superseries = names(res2))
res3$id <- res2
dict_series <- res3 %>% unnest(id) %>% na.omit()
gds2 <- gds %>% left_join(dict_series) %>% 
  mutate(superseries = ifelse(is.na(superseries),
                              id,
                              superseries)) %>% 
  group_by(superseries) %>% 
  summarize(has_meta1 = ifelse(any(has_meta1 == "yes"), "yes", "no"),
            has_meta2 = ifelse(any(has_meta2 == "yes"), "yes", "no"),
            has_object1 = ifelse(any(has_object1 == "yes"), "yes", "no"),
            has_object2 = ifelse(any(has_object2 == "yes"), "yes", "no")) %>% 
  mutate(usable = ifelse(has_meta1 == "yes" | 
                           has_meta2 == "yes" | 
                           has_object1 == "yes" | 
                           has_object2 == "yes",
                         "yes", "no")) %>% 
  mutate(usable2 = ifelse(has_meta1 == "yes" | has_meta2 == "yes", "meta",
                          ifelse(has_object1 == "yes" | has_object2 == "yes", "object", "no")))

gds_fixed <- gds %>% inner_join(gds2 %>% select(id = superseries, fixed_usable = usable, usable2))
gds <- gds_fixed %>% select(-usable) %>% rename(usable = fixed_usable)
```

Merged super and subseries: `r length(res) - nrow(gds)`

```{r fixfalsep}
gds_old <- gds
fp <- read_tsv(here("inst", "extdata", "falsepos.txt"), col_names = F) %>% pull(1)
gds <- gds %>% mutate(usable = ifelse(id %in% fp, "no", usable))
```

The fraction of GEO entries with potential metadata (file with "meta", "annot", or "type" in filename or rda/rds/rdata files) is `r (gds_old$usable == "yes") %>% mean(na.rm = TRUE)`. Note however that these terms include some false positives (such as gene annotation file, patient metadata, and phenotype table), which we manually inspected and corrected (false positive fraction at `r length(fp) / sum(gds$has_meta1 == "yes", na.rm = TRUE)`). Final fraction: `r (gds$usable == "yes") %>% mean(na.rm = TRUE)`

# Parse database from Svensson et al.

```{r curated}
library(googlesheets4)
gs4_deauth()
gs_data <- read_sheet("https://docs.google.com/spreadsheets/d/1En7-UV0k0laDiIfjFkdn7dggyR7jIk3WH8QgXaMOZF0", sheet = "Data") %>% filter(Measurement == "RNA-seq") %>% filter(str_detect(`Data location`, "^GSE"))

gs_ids <- gs_data$`Data location` %>% strsplit(", ") %>% unlist()
```

```{r matchsub}
library(GEOquery)
get_subseries <- function(geo) {
  res <- tryCatch(
    suppressMessages(getGEO(GEO = geo,
                            filename = NULL, destdir = tempdir(), 
                            GSElimits = NULL, GSEMatrix = FALSE, 
                            AnnotGPL = FALSE, getGPL = TRUE,
                            parseCharacteristics = TRUE)),
    error = function(e) {"notavail"})
  if (class(res) != "GSE") {
    return(NA)
  }
  resr <- res@header$relation
  resr[str_detect(resr, "SubSeries of: |SuperSeries of: ")] %>% str_remove("SubSeries of: |SuperSeries of: ")
}

checksub <- purrr::map(
  setdiff(gs_ids, gds$id),
  function(x) {
    temp <- get_subseries(x)
    if (length(temp) == 1) {
      if (is.na(temp)) {
        NA
      }
      else {
        temp#length(intersect(temp, gds$id)) == 0
      }
    } else {
      temp#length(intersect(temp, gds$id)) == 0
    }
  }) 

names(checksub) <- setdiff(gs_ids, gds$id)
checksub_ids <- checksub
checksub <- sapply(checksub,
                   function(x) length(intersect(x, gds$id)) == 0) %>% unlist()

gs_data_processed <- gs_data %>% inner_join(gds, by = c("Data location" = "id"))
```

Overlap of scRNA-seq GEO entries between manual curation and the GEO query is `r 1 - sum(checksub, na.rm = T)/(length(gs_ids) - sum(is.na(checksub)))`. Number of entries not public: `r sum(is.na(checksub))`. Fraction with metadata: `r ((sum(gs_data_processed$usable == "yes") + unlist(checksub_ids) %in% (gds %>% filter(usable == "yes") %>% pull(id)) %>% sum())) / (nrow(gs_data_processed) + unlist(checksub_ids) %in% (gds %>% filter(usable == "yes") %>% pull(id)) %>% sum())`.

# by year

```{r frac}
frac <- gds %>% na.omit() %>% 
  group_by(year, usable) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  pivot_wider(names_from = usable, values_from = n) %>% 
  replace_na(list(no = 0, yes = 0)) %>% 
  mutate(fraction = yes/(yes + no), total = yes + no) %>% 
  mutate(fraction = round(fraction, digits = 3))

g <- ggplot(gds %>% na.omit() %>% filter(year >= 2012),
       aes(x = year, fill = usable)) +
  geom_bar() +
  scale_fill_brewer() +
  #scale_fill_viridis_d(option = "plasma") +
  geom_text(data = frac %>% filter(year >= 2012), color = "black", aes(x = year, y = total + 30, label = fraction, fill = NA), size = 3) +
  cowplot::theme_cowplot() +
  scale_x_continuous(breaks = c(2012:2020)) +
  ylab("total studies") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 14), 
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14))# + ggtitle("scRNAseq datasets by year")
g
#cowplot::save_plot("geo_year.pdf", g, base_asp = 1, base_width = 5)

# ggplot(gds %>% na.omit() %>% filter(year == "2020") %>% 
#          mutate(month = factor(month, levels = c("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"))),
#        aes(x = month, fill = usable)) +
#   geom_bar() +
#   scale_fill_viridis_d(option = "plasma") +
#   cowplot::theme_cowplot() +
#   ggtitle("scRNAseq datasets by month in 2020")
# frac2 <- gds %>% na.omit() %>% 
#   mutate(year = ifelse(month %in% c("Jan", "Feb", "Mar", "Apr", "May", "Jun"), 
#                        year, year + 0.5)) %>% 
#   group_by(year, usable) %>% 
#   summarize(n = n()) %>% 
#   ungroup() %>% 
#   pivot_wider(names_from = usable, values_from = n) %>% 
#   replace_na(list(no = 0, yes = 0)) %>% 
#   mutate(fraction = yes/(yes + no), total = yes + no) %>% 
#   mutate(fraction = round(fraction, digits = 3))
# fit <- lm(fraction ~ year, data = frac2 %>% filter(year >= 2016))
# ggplot(frac2 %>% filter(year >= 2016),
#        aes(x = year, y = fraction, color = total)) +
#   geom_point() +
#   geom_line(data = fortify(fit), aes(x = year, y = .fitted), color = "black") +
#   scale_color_viridis_c(direction = -1) +
#   cowplot::theme_cowplot() +
#   ylab("total studies") +
#   ylim(c(0, 0.25)) +
#   ggtitle("metadata fraction by year")
```

# by journal

```{r journalplot}
# ids_nat <- gs_data %>% filter(Journal == "Nature" | str_detect(Journal, "^Nat "))
# ids_cell <- gs_data %>% filter(Journal %in% c("Cell","Cell Reports", "Developmental Cell", "Cell Stem Cell", "Molecular Cell", "Cancer Cell", "Immunity", "Neuron"))
# ids_sci <- gs_data %>% filter(Journal %in% c("Science", "Sci Rep"))
# ids_elife <- gs_data %>% filter(Journal == "eLife")
# ids_pnas <- gs_data %>% filter(Journal == "Proc Natl Acad Sci USA")
# ids_gb <- gs_data %>% filter(Journal == "Genome Biol")

gds_j <- gds %>%
  mutate(journal_group = case_when(
    journal %in% c("Nature", "Science", "Cell") ~ "CNS",
    journal %in% c("Nat Commun", "Nat Metab", "iScience", "Cell Res", "Sci Rep", "Sci Immunol", "Sci Adv", "Cell Rep", "Dev Cell", "Cell Stem Cell", "Stem Cell Reports", "Mol Cell", "Cancer Cell", "Immunity", "Neuron", "Cell Syst", "Cell Metab", "Cell Host Microbe") ~ "CNS sub",
    str_detect(journal, "^Nat ") ~ "CNS sub",
    journal %in% c("Genome Biol", "Genome Med") ~ "BMC",
    str_detect(journal, "BMC") ~ "BMC",
    journal %in% c("Genes Dev", "Genome Res") ~ "CSHL",
    journal %in% c("Elife") ~ "eLife",
    # journal %in% c("Proc Natl Acad Sci U S A") ~ "PNAS",
    str_detect(journal, "PLoS") ~ "PLOS",
    #journal %in% c("Nucleic Acids Res.") ~ "NAR",
    TRUE ~ "other"
    )) %>% 
  mutate(journal_group2 = case_when(
    journal %in% c("Nature", "Nat Commun", "Nat Metab", "Cell Res.") ~ "Nature",
    str_detect(journal, "^Nat. ") ~ "Nature",
    journal %in% c("Science", "Sci Rep", "Sci Immunol", "Sci Adv") ~ "Science",
    journal %in% c("Cell", "Cell Rep", "Dev. Cell", "Cell Stem Cell", "Stem Cell Reports", "Mol. Cell", "Cancer Cell", "Immunity", "Neuron", "Cell Syst", "Cell Metab.", "Cell Host Microbe", "iScience") ~ "Cell",
    #journal %in% c("Genome Biol.", "Genome Med") ~ "BMC",
    #str_detect(journal, "BMC") ~ "BMC",
    #str_detect(journal, "^Front") ~ "Frontiers",
    #journal %in% c("Genes Dev.", "Genome Res.") ~ "CSHL",
    #str_detect(journal, "PLoS") ~ "PLOS",
    #journal %in% c("Nucleic Acids Res.") ~ "NAR",
    TRUE ~ "other"
    )) %>% 
  filter(journal != "error") %>% 
  na.omit()

gds_j_frac <- gds_j %>% 
  filter(year >= 2017) %>% 
  group_by(journal_group, usable) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  pivot_wider(names_from = usable, values_from = n) %>% 
  replace_na(list(no = 0, yes = 0)) %>% 
  mutate(fraction = yes/(yes + no), total = yes + no) %>% 
  mutate(fraction = round(fraction, digits = 3)) %>% 
  arrange(desc(fraction))

g <- ggplot(gds_j %>% filter(year >= 2017) %>% mutate(journal_group = factor(journal_group, levels = gds_j_frac$journal_group)),
       aes(x = journal_group, fill = usable)) +
  geom_bar() +
  scale_fill_brewer() +
  #scale_fill_viridis_d(option = "plasma") +
  geom_text(data = gds_j_frac, color = "black", aes(x = journal_group, y = total + 30, label = fraction, fill = NA), size = 3) +
  cowplot::theme_cowplot() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 14), 
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) #+ ggtitle("scRNAseq datasets by journal")
g
#cowplot::save_plot("geo_family.pdf", g, base_asp = 1, base_width = 5)
```

```{r journaldot}
gds_sum <- gds_j %>% filter(year >= 2017) %>% 
  group_by(journal, journal_group, journal_group2) %>% 
  summarize(n = n(), frac = sum(usable == "yes") / n) %>% 
  ungroup()

g <- ggplot(gds_sum %>% filter(n >= 3) %>% 
         mutate(journal_group = factor(journal_group, 
                                       # levels = c("Cell", "Nature", "Science", "BMC", "other")
                                       levels = c("CNS", "CNS sub", "BMC", "CSHL" , "eLife", "PLOS", "other")
                                       )),
       aes(x = n, y = frac, color = journal_group)) +
  geom_point() +
  scale_color_brewer(palette = "Accent") +
  cowplot::theme_cowplot() +
  # ggtitle("scRNAseq datasets by journal family") + 
  xlab("total studies (since 2017)") +
  ylab("fraction with metadata") +
  ylim(c(0, 1)) +
  theme(axis.text.x = element_text(size = 14), 
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14)) +
  scale_x_log10()
g
#cowplot::save_plot("geo_journal.pdf", g, base_asp = 1, base_width = 5)
```

```{r getcitations}
# gds_c <- gds_j %>% mutate(jy = str_c(journal, year, sep = ",")) %>% 
#   group_by(journal, year, usable) %>% 
#   filter(n() >= 4) %>% 
#   filter(year <= 2018) %>% 
#   ungroup() %>% 
#   select(id, usable, jy, cite) %>% 
#   filter(!is.na(cite)) %>% 
#   mutate(cite = as.numeric(cite))
# jys <- gds_c %>% select(usable, jy) %>% distinct() %>% group_by(jy) %>% filter(n() == 2) %>% pull(jy)
# g <- ggplot(gds_c %>% filter(jy %in% jys),
#        aes(x = jy, y = cite)) +
#   geom_boxplot(aes(fill = usable), outlier.shape = NA) +
#   cowplot::theme_cowplot() +
#   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
#   #coord_cartesian(ylim = c(0,500)) +
#   ggtitle("citations binned by year + journal + usable")
# g
```

# by IF

```{r if}
# ifs <- scholar::get_impactfactor(sapply(gds_j$pubmed, function(x) str_remove(str_remove(x$journal[1], "^The |^the |^THE "), " \\(.+")))
# ifs_abbr <- scholar::get_impactfactor(gds_j$journal)
# cut <- ifs_abbr %>% filter(Journal %in% c("CELL", "NATURE", "SCIENCE")) %>% pull(ImpactFactor) %>% max()
# ifs2 <- bind_cols(if_long = ifs$ImpactFactor, if_short = ifs_abbr$ImpactFactor) %>% 
#   mutate(if_long = ifelse(if_long > cut, NA, if_long),
#          if_short = ifelse(if_short > cut, NA, if_short)) %>% 
#   mutate(ifs = pmax(if_long, if_short, na.rm = TRUE))
# gds_j$ifs <- ifs2$ifs
# 
# checked_j <- gds_j %>% group_by(journal, ifs) %>% summarise(n = n()) %>%
#   mutate(ifs = case_when(
#     journal == "JCI Insight" ~ 6.205,
#     journal == "Science Immunology" ~ 8.160,
#     journal == "J Am Soc Nephrol" ~ 9.274,
#     journal == "Genes Dev" ~ 9.527,
#     journal == "Blood Adv" ~ 4.584,
#     journal == "Curr Biol" ~ 5.720,
#     journal == "Protein Cell" ~ 10.164,
#     journal == "iScience" ~ 4.447,
#     journal == "Nat Struct Mol Biol" ~ 11.98,
#     TRUE ~ ifs
#   ))
# 
# write_csv(checked_j, here("inst", "extdata", "ifs.csv"))
checked_j <- read_csv(here("inst", "extdata", "ifs.csv")) %>% select(-n)
gds_j <- gds_j %>% left_join(checked_j)
# gds_c_if <- gds_j %>% 
#   filter(ifs < 51.941) %>% 
#   mutate(bin = case_when(
#     ifs < 5 ~ "5",
#     ifs < 10 ~ "10",
#     ifs < 30 ~ "30",
#     TRUE ~ "+",
#   )) %>% 
#   filter(year == 2018 | year == 2017) %>% 
#   select(pubmed, usable, journal, cite, bin, year) %>% 
#   distinct() %>% 
#   filter(!is.na(cite)) %>% 
#   mutate(cite = as.numeric(cite)) %>% 
#   mutate(bin = factor(bin, levels = c("5", "10", "30", "+")))

# g <- ggplot(gds_c_if,
#        aes(x = bin, y = cite)) +
#   #facet_grid(.~year) +
#   geom_boxplot(aes(fill = usable), outlier.shape = NA) +
#   cowplot::theme_cowplot() +
#   #theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
#   coord_cartesian(ylim = c(0,500)) +
#   ggtitle("citations binned by IF, 2018")
# g
```

```{r citations}
gds_c_if2 <- gds_j %>% 
  mutate(bin = case_when(
    ifs < 5 ~ "5",
    ifs < 10 ~ "10",
    ifs < 30 ~ "30",
    TRUE ~ "+",
  )) %>% 
  filter(year == 2017 | year == 2018) %>% #| year == 2016| year == 2019) %>% 
  # mutate(year = ifelse(month %in% c("Jan", "Feb", "Mar", "Apr", "May", "Jun"),
  #                      year + 0.5,
  #                      year)) %>%
  select(pubmed, usable, journal, cite, bin, year, ifs) %>% 
  distinct() %>% 
  filter(!is.na(cite)) %>% 
  mutate(cite = as.numeric(cite)) %>% 
  mutate(bin = factor(bin, levels = c("5", "10", "30", "+")))

g <- ggplot(gds_c_if2,
            aes(x = ifs, y = cite, group = usable, color = usable, shape = usable)) +
  facet_grid(.~year) +
  geom_point(alpha = 0.87) +
  geom_point(alpha = 0.87, 
             data = gds_c_if2 %>% filter(usable == "yes")) +
  cowplot::theme_cowplot() +
  #scale_y_log10() +
  geom_smooth(method = "lm", fill = NA) +
  #scale_color_viridis_d(option = "a") +
  scale_color_brewer() +
  #theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  coord_cartesian(ylim = c(0,500)) +
  xlab("journal IF") + ylab("# of citations") #+ ggtitle("citations by IF, 2017/8")
g
#cowplot::save_plot("geo_cite.pdf", g, base_asp = 1, base_width = 5)
# 
# g2 <- ggplot(gds_c_if2,
#             aes(x = ifs, y = cite, group = usable, color = usable, shape = usable)) +
#   facet_grid(.~year) +
#   geom_point(alpha = 0.76) +
#   cowplot::theme_cowplot() +
#   scale_y_log10() +
#   geom_smooth(method = "lm", fill = NA) +
#   scale_color_viridis_d(option = "plasma") +
#   #scale_color_brewer() +
#   #theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
#   #coord_cartesian(ylim = c(0,500)) +
#   xlab("journal IF") + ylab("# of citations") #+ ggtitle("citations by IF, 2017/8")
# g2
```

```{r authors, eval=F}
get_authors <- function(pubmed) {
  tryCatch(pubmed[[1]] %>% mutate(fullname = str_c(firstname, lastname, sep = "_")) %>%
             pull(fullname),
           error = function(e) {"NA"})
}
get_firauthors <- function(pubmed) {
  tryCatch(pubmed %>% mutate(fullname = str_c(firstname, lastname, sep = "_")) %>% 
             #dplyr::slice(n()) %>%
             dplyr::slice(1) %>%
             pull(fullname),
           error = function(e) {"NA"})
}
get_corauthors <- function(pubmed) {
  tryCatch(pubmed[[1]] %>% filter(!is.na(email)) %>% mutate(fullname = str_c(firstname, lastname, sep = "_")) %>% pull(fullname),
           error = function(e) {"NA"})
}
gds_a <- gds %>% select(id, journal, usable, year, pubmed)

fir <- list()
for (i in 1:nrow(gds_a)) {
  fir[[i]] <- get_firauthors(gds$pubmed[i])
}
firs <- fir %>% unlist() %>% unique()

cor <- list()
for (i in 1:nrow(gds_a)) {
  cor[[i]] <- get_corauthors(gds$pubmed[i])
}
cors <- cor %>% unlist() %>% unique()

authors <- list()
for (i in 1:nrow(gds_a)) {
  authors[[i]] <- get_authors(gds$pubmed[i])
}
gds_a$authors <- authors

gds_as <- gds_a %>% select(-pubmed) %>%
  filter(year >= 2017) %>%
  unnest("authors") %>%
  group_by(authors) %>%
  summarize(n = n(), yes = sum(usable == "yes")) %>%
  #filter(authors %in% cors) %>% 
  mutate(frac = yes/n) %>% 
  mutate(author_type = ifelse(authors %in% firsc_check, "yes", "no"))

ggplot(gds_as %>% filter(authors != "NA") %>% filter(n >= 3),
       aes(x = n, y = frac, color = author_type)) +
  facet_grid(.~author_type) +
  geom_point() +
  scale_color_brewer() +
  cowplot::theme_cowplot() +
  ggtitle("scRNAseq datasets by author") +
  xlab("total") +
  ylim(c(0, 1)) +
  scale_x_log10()
```

```{r, eval=F}
authors <- list()
for (i in 1:nrow(gds_a)) {
  authors[[i]] <- get_authors(gds$pubmed[i])
}
gds_a$authors <- authors

class <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/categories-idx.tsv") %>% filter(Category %in% c("Classification")) %>% pull(Tool)
doi <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/doi-idx.tsv") %>% filter(Tool %in% class) %>% pull(DOI)
tools <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/references.tsv") %>% pull(Title) %>%
  str_remove_all(":|,|\\(|\\)") %>%
  str_replace_all("’s | or | and |–|—a |—", " ") %>%
  str_remove_all(" Fig\\. 1\\.") %>% 
  na.omit()
classy <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/references.tsv") %>% filter(!is.na(Title)) %>% pull(DOI) %in% doi

# get pubmed ids for tools
tau <- list()
for (i in 1:length(tools)) {
  # if (i%%10 == 0) {
  #   closeAllConnections()
  #   print(i)
  # }
  tau[[i]] <- easyPubMed::get_pubmed_ids_by_fulltitle(tools[i],
                                                      api_key = "9cc836a53f5c509a6783283e694da4fb0d08")$IdList$Id
}

# for missing entries, try doi search
try2 <- (read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/references.tsv") %>% pull(DOI))[sapply(tau, is.null)]
try2l <- list()
splitst <- split(try2, ceiling(seq_along(try2)/180))
for (x in splitst) {
  try2l <- c(try2l, rcrossref::id_converter(x)$records$pmid %>% na.omit())
}
try2v <- try2l %>% unlist() %>% unique()

pubmeds_t <- list()
splits <- split(c(tau %>% unlist() %>% na.omit(), try2v) %>% unique() ,
                ceiling(seq_along(c(tau %>% unlist() %>% na.omit(), try2v) %>% unique())/20))
for (i in 1:length(splits)) {
  s <- splits[[i]]
  # if (i%%2 == 0) {
  #   closeAllConnections()
  #   print(i)
  # }
  query_string <- str_c(s, collapse = ",")
  pubmed_file <- easyPubMed::get_pubmed_ids(query_string,
                                api_key = "9cc836a53f5c509a6783283e694da4fb0d08")
  pubmed <- pubmed_file %>% fetch_pubmed_data(retmax = 9999) %>%
    articles_to_list()
  pubmeds1 <- sapply(pubmed, article_to_df, simplify = FALSE)
  names(pubmeds1) <- pubmed_file$IdList
  pubmeds_t <- append(pubmeds_t, pubmeds1)
}

fir <- list()
for (i in 1:length(pubmeds_t)) {
  fir[[i]] <- get_firauthors(pubmeds_t[[i]])
}
firs <- fir %>% unlist() %>% unique()
firs <- setdiff(firs, c("NA"))

pubmeds_tc <- pubmeds_t[tau[classy] %>% unlist()]
firc <- list()
for (i in 1:length(pubmeds_tc)) {
  firc[[i]] <- get_firauthors(pubmeds_tc[[i]])
}
firsc <- firc %>% unlist() %>% unique()
firsc <- setdiff(firsc, c("NA"))

fir_id <- gds_a %>% select(-pubmed) %>%
  #filter(year >= 2017) %>%
  unnest("authors") %>%
  filter(authors %in% firs) %>%
  pull(id) %>% 
  unique()

fir_idc <- gds_a %>% select(-pubmed) %>%
  #filter(year >= 2017) %>%
  unnest("authors") %>%
  filter(authors %in% firsc) %>%
  pull(id) %>% 
  unique()

gds_at <- gds %>% 
  filter(year >= 2017) %>% 
  mutate(tool = case_when(
    id %in% fir_idc ~ "yes",
    # id %in% fir_id ~ "tool",
    TRUE ~ "no"
  )) %>% 
  group_by(tool, usable2) %>% 
  summarize(n = n()) %>% 
  group_by(tool) %>% 
  mutate(fraction = n/sum(n))

frac3 <- gds_at %>% filter(usable2 == "no") %>% 
  mutate(fraction = 1 - fraction) %>% 
  mutate(fraction = round(fraction, digits = 3))

g <- ggplot(gds_at %>% mutate(usable2 = factor(usable2, levels = c("no", "meta", "object"))) %>% 
              mutate(tool = factor(tool, levels = c("no", "yes"))),
       aes(x = tool, fill = usable2, y = fraction)) +
  geom_col() +
  scale_fill_brewer(name = "usable") +
  geom_text(data = frac3, color = "black", aes(y = 1.03, label = fraction), size = 3) +
  cowplot::theme_cowplot() +
  ylab("fraction") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 14), 
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14))
# cowplot::save_plot("geo_authors.pdf", g, base_asp = 1, base_width = 5)
g + ggtitle("scRNAseq datasets by year")
```

```{r}
library(rbiorxiv)
biorxiv_conv <- function(doi1) {
  tryCatch(biorxiv_content(doi = doi1)[[1]]$published,
           error = function(e) doi1)
}

classes <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/categories-idx.tsv")
classes <- split(classes$Tool, classes$Category)

d <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/doi-idx.tsv")
ref <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/references.tsv")

classes <- sapply(classes, function(x) {
  d %>% filter(Tool %in% x) %>% pull(DOI)
})
classes$all <- unlist(classes) %>% unique()
ts <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/references.tsv") %>% filter(!is.na(Title)) %>% 
  mutate(DOI_pub = mapply(biorxiv_conv, DOI))
try2l <- list()

id_conv <- function(doi) {
  tryCatch(rcrossref::id_converter(doi)$records$pmid,
           error = function(e) NA)
}
try2l <- ts %>% filter(DOI_pub != "NA") %>% 
  select(DOI_pub) %>% 
  mutate(pmid = mapply(id_conv, DOI_pub))
# splitst <- split(ts %>% filter(DOI_pub != "NA") %>% pull(DOI_pub), 
#                  ceiling(seq_along(ts %>% filter(DOI_pub != "NA") %>% pull(DOI_pub))/180))
# for (x in splitst) {
#   try2l <- bind_rows(try2l, rcrossref::id_converter(x)$records %>% select(doi, pmid))
# }

tools <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/references.tsv") %>% pull(Title) %>%
  str_remove_all(":|,|\\(|\\)") %>%
  str_replace_all("’s | or | and |–|—a |—", " ") %>%
  str_remove_all(" Fig\\. 1\\.") %>% 
  na.omit()

tau <- list()
for (i in 1:length(tools)) {
  # if (i%%10 == 0) {
  #   closeAllConnections()
  #   print(i)
  # }
  tau[[i]] <- easyPubMed::get_pubmed_ids_by_fulltitle(tools[i],
                                                      api_key = "9cc836a53f5c509a6783283e694da4fb0d08")$IdList$Id
}

tau2 <- tau
tau2[sapply(tau2, is.null)] <- NA
ts$tau <- unlist(tau2)
ts2 <- ts %>% left_join(try2l, by = "DOI_pub") %>%
  mutate(pubmed = ifelse(is.na(pmid), tau, pmid)) %>% 
  select(DOI, pubmed) %>% 
  distinct() %>% 
  column_to_rownames("DOI")
ts2 <- ts2[ts$DOI,]
names(ts2) <- ts$DOI

pubmeds_t2 <- list()
ts2[sapply(ts2, is.null)] <- NA
ts2 <- ts2 %>% unlist()
splits <- split(ts2 %>% na.omit() %>% unique(),
                ceiling(seq_along(ts2 %>% na.omit() %>% unique())/20))
for (i in 1:length(splits)) {
  s <- splits[[i]]
  # if (i%%2 == 0) {
  #   closeAllConnections()
  #   print(i)
  # }
  query_string <- str_c(s, collapse = ",")
  pubmed_file <- easyPubMed::get_pubmed_ids(query_string,
                                api_key = "9cc836a53f5c509a6783283e694da4fb0d08")
  pubmed <- pubmed_file %>% easyPubMed::fetch_pubmed_data(retmax = 9999) %>%
    easyPubMed::articles_to_list()
  pubmeds1 <- sapply(pubmed, easyPubMed::article_to_df, simplify = FALSE)
  names(pubmeds1) <- pubmed_file$IdList
  pubmeds_t2 <- append(pubmeds_t2, pubmeds1)
}

notold <- sapply(pubmeds_t2, function(x) {
  as.numeric(x$year[1])
}, simplify = TRUE) %>% as.data.frame() %>% 
  rownames_to_column("id") %>% 
  filter(`.` >= 2017) %>% 
  pull(id)

pubmeds_t3 <- pubmeds_t2[notold]

get_authors <- function(pubmed) {
  tryCatch(pubmed %>% mutate(fullname = str_c(firstname, lastname, sep = "_")) %>% 
             pull(fullname),
           error = function(e) {"NA"})
}

gds_a <- gds %>% select(id, journal, usable, year, pubmed)
authors <- list()
for (i in 1:nrow(gds_a)) {
  authors[[i]] <- get_authors(gds$pubmed[i][[1]])
}
gds_a$authors <- authors

plot_author <- function(classy) {
  fir <- list()
  for (i in 1:length(pubmeds_t3)) {
    fir[[i]] <- get_authors(pubmeds_t3[[i]])
  }
  firs <- fir %>% unlist() %>% unique()
  firs <- setdiff(firs, c("NA"))
  
  pubmeds_tc <- pubmeds_t3[ts2[classy] %>% na.omit()]
  print(length(pubmeds_tc))
  firc <- list()
  for (i in 1:length(pubmeds_tc)) {
    firc[[i]] <- get_authors(pubmeds_tc[[i]])
  }
  firsc <- firc %>% unlist() %>% unique()
  firsc <- setdiff(firsc, c("NA"))
  firsc_check <<- firsc
  
  fir_id <- gds_a %>% select(-pubmed) %>%
    #filter(year >= 2017) %>%
    unnest("authors") %>%
    group_by(id) %>% 
    #slice(1,2) %>% 
    filter(authors %in% firs) %>%
    pull(id) %>% 
    unique()
  
  fir_idc <- gds_a %>% select(-pubmed) %>%
    #filter(year >= 2017) %>%
    unnest("authors") %>%
    group_by(id) %>% 
    #slice(1,2) %>% 
    filter(authors %in% firsc) %>%
    pull(id) %>% 
    unique()
  
  fir_idc_check <<- fir_idc
  
  gds_at <- gds %>% 
    filter(year >= 2017) %>% 
    mutate(tool = case_when(
      id %in% fir_idc ~ "yes",
      # id %in% fir_id ~ "tool",
      TRUE ~ "no"
    )) %>% 
    group_by(tool, usable2) %>% 
    summarize(n = n()) %>% 
    group_by(tool) %>% 
    mutate(fraction = n/sum(n))
  
  frac3 <- gds_at %>% filter(usable2 == "no") %>% 
    mutate(fraction = 1 - fraction) %>% 
    mutate(fraction = round(fraction, digits = 3))
  
  g <- ggplot(gds_at %>% mutate(usable2 = factor(usable2, levels = c("no", "meta", "object"))) %>% 
                mutate(tool = factor(tool, levels = c("no", "yes"))),
         aes(x = tool, fill = usable2, y = fraction)) +
    geom_col() +
    scale_fill_brewer(name = "usable") +
    geom_text(data = frac3, color = "black", aes(y = 1.03, label = fraction), size = 3) +
    cowplot::theme_cowplot() +
    ylab("fraction") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 14), 
          axis.text.y = element_text(size = 14),
          axis.title.x = element_text(size = 14),
          axis.title.y = element_text(size = 14))
  g
  # cowplot::save_plot("geo_authors.pdf", g, base_asp = 1, base_width = 5)
}
ps <- plot_author(classes$all)
# g <- cowplot::plot_grid(plotlist = ps, labels = str_c(names(ps), sapply(classes, length, simplify = FALSE), sep = ":"))
# 
# cowplot::save_plot("all_tools_save_2017.pdf", g, base_asp = 1, base_width = 35, base_height = 35)
```

# by tool dev

```{r eval = F}
gds <- gds %>% mutate(tool = case_when(
      id %in% fir_idc_check ~ "yes",
      TRUE ~ "no"
    ))
frac2 <- gds %>% na.omit() %>% 
  #mutate(year = ifelse(month %in% c("Jan", "Feb", "Mar", "Apr", "May", "Jun"), year, year + 0.5)) %>% 
  group_by(year, usable2, tool) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  pivot_wider(names_from = usable2, values_from = n) %>% 
  replace_na(list(no = 0, meta = 0, object = 0)) %>% 
  mutate(fraction = (meta + object) / (meta + object + no), total = meta + object + no) %>% 
  mutate(fraction = round(fraction, digits = 3))

# g <- ggplot(gds %>% na.omit() %>% 
#               filter(year >= 2012) %>%
#               mutate(usable2 = factor(usable2, levels = c("no", "meta", "object"))),
#        aes(x = year)) +
#   facet_grid(.~tool) +
#   geom_bar(aes(fill = usable2)) +
#   scale_fill_brewer(name = "usable") +
#   #scale_fill_viridis_d(option = "plasma") +
#   geom_text(data = frac2 %>% filter(year >= 2012), color = "black", aes(x = year, y = total + max(frac2$total) * 0.03, label = fraction), size = 3) +
#   cowplot::theme_cowplot() +
#   scale_x_continuous(breaks = c(2012:2020)) +
#   ylab("total studies") +
#   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 14), 
#         axis.text.y = element_text(size = 14),
#         axis.title.x = element_text(size = 14),
#         axis.title.y = element_text(size = 14))

g <- ggplot(frac2 %>% filter(year >= 2017), aes(x = year, y = fraction, color = tool)) +
  geom_line() +
  geom_point() +
  scale_color_brewer(palette = "Accent") +
  cowplot::theme_cowplot() +
  xlab("year") +
  ylab("fraction with metadata") +
  ylim(c(0, 1)) +
  theme(axis.text.x = element_text(size = 14), 
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14))
# cowplot::save_plot("geo_tool110520.pdf", g, base_asp = 1, base_width = 5)
g
```

# reanalysis
```{r, eval = F}
checks_re <- suppressWarnings(map(gds$geo,
                                  function(x) str_detect(str_to_lower(x), 
                                                         "novel") %>% any())) #"re-analy|reanaly|publicly available"
gds$re <- ifelse(unlist(checks_re), "yes", "no")
```

```{r, eval = F}
library(tidytext)
summaries <- suppressWarnings(map(gds %>% filter(year >= 2017) %>% pull(geo),
                                  function(x) str_c(x$summary, collapse = " ")))
gds_text <- gds %>% filter(year >= 2017) %>% select(id, usable2) %>% mutate(usable2 = ifelse(usable2 == "no", "no", "yes"))
gds_text$text <- unlist(summaries)
gds_text2 <- unnest_tokens(gds_text, "word", "text") %>% 
  anti_join(stop_words) %>% 
  #mutate(word = str_extract(word, "[:alpha:]+")) %>% 
  count(id, word, sort = TRUE) %>%
  add_count(id, wt = n, name = "total_word") %>% 
  mutate(proportion = n / total_word) %>% 
  select(-total_word, -n) %>% 
  pivot_wider(names_from = id, values_from = proportion, 
              values_fill = list(proportion = 0)) %>% 
  na.omit() %>% 
  column_to_rownames("word") %>% 
  t()
gds_text2[gds_text2 > 0] <- 1
gds_text3 <- gds_text2 %>% as.data.frame() %>%
  rownames_to_column("geoid") %>% 
  left_join(gds_text %>% select(id, usable2), by = c("geoid" = "id"))
data <- gds_text3 %>% select(usable2, everything(), -geoid)
data2 <- data %>% group_by(usable2) %>% 
  summarize_all(mean)
data3 <- data2 %>% column_to_rownames("usable2") %>% 
  t() %>%
  as.data.frame() %>% 
  rownames_to_column("word") %>% 
  mutate(diff = yes - no, fc = log2(yes/no)) %>% 
  filter(yes >= 0.1)
```

```{r}
gds$upload <- sapply(gds$geo, function(x) {
  temp <- x$contact_name
  if (is.null(temp)) {
    return("NA")
  } else {
    temp
  }})

gds_cor <- gds_a %>% unnest(authors) %>% 
  group_by(id) %>% 
  slice(n())

gds_up <- gds %>% select(id, usable2, upload, year) %>% 
  separate(upload, ",", into = c("F", "M", "L")) %>% 
  mutate(full1 = str_c(str_to_title(F), " ", str_to_upper(M), "_", str_to_title(L))) %>% 
  mutate(M = str_sub(M, 1, 1)) %>% 
  mutate(full2 = str_c(str_to_title(F), " ", str_to_upper(M), "_", str_to_title(L))) %>% 
  mutate(full3 = str_c(str_to_title(F), "_", str_to_title(L))) %>% 
  mutate(check = case_when(
    (full1 %in% firsc_check) | (full2 %in% firsc_check) |(full3 %in% firsc_check) ~ "upload", 
    id %in% fir_idc_check ~ "author",
    TRUE ~ "no"
    )) %>% 
  left_join(gds_cor %>% select(id, authors)) #%>% filter(full1 != authors, full2 != authors, full3 != authors)

gds_up %>% group_by(check, usable2) %>% 
  summarize(n = n()) %>% 
  group_by(check) %>% 
  mutate(frac = n/sum(n))

frac4 <- gds_up %>% na.omit() %>% 
  #mutate(year = ifelse(month %in% c("Jan", "Feb", "Mar", "Apr", "May", "Jun"), year, year + 0.5)) %>% 
  group_by(year, usable2, check) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  pivot_wider(names_from = usable2, values_from = n) %>% 
  replace_na(list(no = 0, meta = 0, object = 0)) %>% 
  mutate(fraction = (meta + object) / (meta + object + no), total = meta + object + no) %>% 
  mutate(fraction = round(fraction, digits = 3)) %>% 
  mutate(check = factor(check, levels = c("upload", "author", "no")))

g <- ggplot(frac4 %>% filter(year >= 2017), aes(x = year, y = fraction, color = check)) +
  geom_line() +
  geom_point() +
  scale_color_brewer(palette = "Accent") +
  cowplot::theme_cowplot() +
  xlab("year") +
  ylab("fraction with metadata") +
  ylim(c(0, 1)) +
  theme(axis.text.x = element_text(size = 14), 
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14))
# cowplot::save_plot("geo_tool110620.pdf", g, base_asp = 1, base_width = 5)
g
```

