---
title: 'Assessment of missing cell-level metadata in single cell GEO records'
date: '`r Sys.Date()`'
# package: someta
author:
  - name: Rui Fu
    affiliation: New York Genome Center
output:
  BiocStyle::html_document:
    toc_float: true

vignette: >
    %\VignetteIndexEntry{metadata_report}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}  
---

```{r "knitr options", echo = FALSE, message = FALSE, warning = FALSE}
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  echo = FALSE,
  collapse = TRUE,
  fig.align = "center",
  comment = "#>",
  crop = NULL
)
```

```{r load}
library(here)
library(tidyverse)
date <- format(format = "%m%d%y", Sys.Date())
```

# Current GEO query used: 

```
"expression profiling by high throughput sequencing" AND
("single nuclei" OR "single cell" OR "scRNAseq" OR "scRNA-seq" OR "snRNAseq" OR "snRNA-seq")
```

```{r eutil, eval = T}
library(httr)
print(date)
esearch <- "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=gds&term=%22expression%20profiling%20by%20high%20throughput%20sequencing%22%20AND%20(%22single%20nuclei%22%20OR%20%22single%20cell%22%20OR%20%22scRNAseq%22%20OR%20%22scRNA-seq%22%20OR%20%22snRNAseq%22%20OR%20%22snRNA-seq%22)&usehistory=y&retmode=json"
get1 <- GET(esearch)
webenv <- content(get1)$esearchresult$webenv
efetch <- paste0(
  "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=gds&query_key=1&WebEnv=",
  webenv
)

if (!(file.exists(here("inst", "extdata", date, paste0("gds_result_", date, ".txt"))))) {
  dir.create(here("inst", "extdata", date), showWarnings = FALSE)
  GET(
    efetch,
    write_disk(here("inst", "extdata", date, paste0("gds_result_", date, ".txt")))
  )
}
```

# Parse GEO query

```{r entries}
# if (format(format = "%m%d%y", file.info(here("inst", "extdata", "current_geo.rds"))$ctime) == date) {
if (!(file.exists(here("inst", "extdata", date, paste0("gds_result_", date, ".txt"))))) {
  gds <- readRDS(here("inst", "extdata", "current_geo.rds")) %>%
    mutate(year = as.numeric(year))
  gds %>% saveRDS(here("inst", "extdata", date, paste0("geo_", date, ".rds")))
} else {
  if (file.exists(here("inst", "extdata", date, paste0("geo_", date, ".rds")))) {
    gds <- readRDS(here("inst", "extdata", date, paste0("geo_", date, ".rds"))) %>%
      mutate(year = as.numeric(year))
  } else {
    source(here("utils", "geo_query.R"))
    gds <- readRDS(here("inst", "extdata", date, paste0("geo_", date, ".rds"))) %>%
      mutate(year = as.numeric(year))
  }
}

gds_backup <- readRDS(here("inst", "extdata", "101520", paste0("geo_", "101520", ".rds"))) %>%
  mutate(year = as.numeric(year))
gds_orig <- gds
```

```{r filteragain}
checks <- suppressWarnings(map(
  gds$geo,
  function(x) {
    str_detect(
      str_to_lower(x),
      "single nuclei|single-nucleus|single cell|single-cell|scrnaseq|scrna-seq|snrnaseq|snrna-seq"
    ) %>% any()
  }
))
gds <- gds[unlist(checks), ]
```

Total entries: `r nrow(gds_orig)`

Number of entries filtered out because key words were not found: `r sum(!unlist(checks), na.rm = TRUE)`

```{r fixseries}
parse_subseries <- function(geolist) {
  tryCatch(
    {
      rel <- geolist$relation
      rel[str_detect(rel, "^SuperSeries of: ")] %>%
        str_remove("^SuperSeries of: ")
    },
    error = function(e) {
      return(NA)
    }
  )
}

res <- map(
  gds$geo,
  parse_subseries
)
names(res) <- gds$id
res2 <- res[lengths(res) != 0]
res2 <- map(
  res2,
  function(x) ifelse(x %in% gds$id, x, NA)
)
res3 <- data.frame(superseries = names(res2))
res3$id <- res2
dict_series <- res3 %>%
  unnest(id) %>%
  na.omit()
gds2 <- gds %>%
  left_join(dict_series) %>%
  mutate(superseries = ifelse(is.na(superseries),
    id,
    superseries
  )) %>%
  group_by(superseries) %>%
  summarize(
    has_meta1 = ifelse(any(has_meta1 == "yes"), "yes", "no"),
    has_meta2 = ifelse(any(has_meta2 == "yes"), "yes", "no"),
    has_object1 = ifelse(any(has_object1 == "yes"), "yes", "no"),
    has_object2 = ifelse(any(has_object2 == "yes"), "yes", "no")
  ) %>%
  mutate(usable = ifelse(has_meta1 == "yes" |
    has_meta2 == "yes" |
    has_object1 == "yes" |
    has_object2 == "yes",
  "yes", "no"
  )) %>%
  mutate(usable2 = ifelse(has_meta1 == "yes" | has_meta2 == "yes", "meta",
    ifelse(has_object1 == "yes" | has_object2 == "yes", "object", "no")
  ))

gds_fixed <- gds %>% inner_join(gds2 %>% select(id = superseries, fixed_usable = usable, usable2))
gds <- gds_fixed %>%
  select(-usable) %>%
  rename(usable = fixed_usable)
```

Merged super and subseries: `r length(res) - nrow(gds)`

```{r fixfalsep}
gds_old <- gds
fp <- read_tsv(here("inst", "extdata", "falsepos.txt"), col_names = F) %>% pull(1)
gds <- gds %>% mutate(usable = ifelse(id %in% fp, "no", usable))
```

The fraction of GEO entries with potential metadata (file with "meta", "annot", "clustering", "colData", or "type" in filename or rda/rds/rdata/h5ad/loom files) is `r (gds_old$usable == "yes") %>% mean(na.rm = TRUE)`. Note however that these terms include some false positives (such as gene annotation file, patient metadata, and phenotype table), which we manually inspected and corrected (false positive fraction at `r length(fp) / sum(gds$has_meta1 == "yes", na.rm = TRUE)`). Final fraction: `r (gds$usable == "yes") %>% mean(na.rm = TRUE)`

# Comparison of parsed data to database from Svensson et al.

```{r curated}
library(googlesheets4)
gs4_deauth()
gs_data <- read_sheet("https://docs.google.com/spreadsheets/d/1En7-UV0k0laDiIfjFkdn7dggyR7jIk3WH8QgXaMOZF0", sheet = "Data") %>%
  filter(Measurement == "RNA-seq") %>%
  filter(str_detect(`Data location`, "^GSE"))

gs_ids <- gs_data$`Data location` %>%
  strsplit(", ") %>%
  unlist()
```

```{r matchsub}
library(GEOquery)
get_subseries <- function(geo) {
  res <- tryCatch(
    suppressMessages(getGEO(
      GEO = geo,
      filename = NULL, destdir = tempdir(),
      GSElimits = NULL, GSEMatrix = FALSE,
      AnnotGPL = FALSE, getGPL = TRUE,
      parseCharacteristics = TRUE
    )),
    error = function(e) {
      "notavail"
    }
  )
  if (class(res) != "GSE") {
    return(NA)
  }
  resr <- res@header$relation
  resr[str_detect(resr, "SubSeries of: |SuperSeries of: ")] %>% str_remove("SubSeries of: |SuperSeries of: ")
}

checksub <- purrr::map(
  setdiff(gs_ids, gds$id),
  function(x) {
    temp <- get_subseries(x)
    if (length(temp) == 1) {
      if (is.na(temp)) {
        NA
      }
      else {
        temp
      }
    } else {
      temp
    }
  }
)

names(checksub) <- setdiff(gs_ids, gds$id)
checksub_ids <- checksub
checksub <- sapply(
  checksub,
  function(x) length(intersect(x, gds$id)) == 0
) %>% unlist()

gs_data_processed <- gs_data %>% inner_join(gds, by = c("Data location" = "id"))
gds_shared <- gs_data_processed %>% select(id = `Data location`, usable) %>% 
  bind_rows(data.frame(id = names(checksub)[checksub], usable = NA))

# write_tsv(gds_shared, "geo_vs_Svensson.tsv.gz")
```

Overlap of scRNA-seq GEO entries between manual curation and the GEO query is `r 1 - sum(checksub, na.rm = T)/(length(gs_ids) - sum(is.na(checksub)))`. Number of entries not public: `r sum(is.na(checksub))`. Fraction with metadata: `r ((sum(gs_data_processed$usable == "yes") + unlist(checksub_ids) %in% (gds %>% filter(usable == "yes") %>% pull(id)) %>% sum())) / (nrow(gs_data_processed) + unlist(checksub_ids) %in% (gds %>% filter(usable == "yes") %>% pull(id)) %>% sum())`.

# GEO cell-level metadata availability by year

```{r frac}
frac <- gds %>%
  na.omit() %>%
  group_by(year, usable) %>%
  summarize(n = n()) %>%
  ungroup() %>%
  pivot_wider(names_from = usable, values_from = n) %>%
  replace_na(list(no = 0, yes = 0)) %>%
  mutate(fraction = yes / (yes + no), total = yes + no) %>%
  mutate(fraction = round(fraction, digits = 3))

g <- ggplot(
  gds %>% na.omit() %>% filter(year >= 2012) %>% mutate(usable2 = factor(usable2, levels = c("no", "meta", "object"))),
  aes(x = year)
) +
  geom_bar(aes(fill = usable2)) +
  scale_fill_brewer(name = "usable") +
  geom_text(data = frac %>% filter(year >= 2012), color = "black", aes(x = year, y = total + 30, label = fraction), size = 3) +
  cowplot::theme_cowplot() +
  scale_x_continuous(breaks = c(2012:format(Sys.time(), "%Y") %>% as.numeric())) +
  ylab("total studies") +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 14),
    axis.text.y = element_text(size = 14),
    axis.title.x = element_text(size = 14),
    axis.title.y = element_text(size = 14)
  )
g # fig2a

gds_a <- gds %>% 
  na.omit() %>% 
  filter(year >= 2012) %>% 
  mutate(usable2 = factor(usable2, levels = c("no", "meta", "object"))) %>% 
  select(id, year, usable, usable2)

# write_tsv(gds_a, "fig2a.tsv.gz")
```

# GEO cell-level metadata availability by journal

```{r journalplot}
gds_j <- gds %>%
  mutate(journal_group = case_when(
    journal %in% c("Nature", "Science", "Cell") ~ "CNS",
    journal %in% c("Nat Commun", "Nat Metab", "iScience", "Cell Res", "Sci Rep", "Sci Immunol", "Sci Adv", "Cell Rep", "Dev Cell", "Cell Stem Cell", "Stem Cell Reports", "Mol Cell", "Cancer Cell", "Immunity", "Neuron", "Cell Syst", "Cell Metab", "Cell Host Microbe") ~ "CNS sub",
    str_detect(journal, "^Nat ") ~ "CNS sub",
    journal %in% c("Genome Biol", "Genome Med") ~ "BMC",
    str_detect(journal, "BMC") ~ "BMC",
    journal %in% c("Genes Dev", "Genome Res") ~ "CSHL",
    journal %in% c("Elife") ~ "eLife",
    str_detect(journal, "PLoS") ~ "PLOS",
    TRUE ~ "other"
  )) %>%
  mutate(journal_group2 = case_when(
    journal %in% c("Nature", "Nat Commun", "Nat Metab", "Cell Res.") ~ "Nature",
    str_detect(journal, "^Nat. ") ~ "Nature",
    journal %in% c("Science", "Sci Rep", "Sci Immunol", "Sci Adv") ~ "Science",
    journal %in% c("Cell", "Cell Rep", "Dev. Cell", "Cell Stem Cell", "Stem Cell Reports", "Mol. Cell", "Cancer Cell", "Immunity", "Neuron", "Cell Syst", "Cell Metab.", "Cell Host Microbe", "iScience") ~ "Cell",
    TRUE ~ "other"
  )) %>%
  filter(journal != "error") %>%
  na.omit()

gds_j_frac <- gds_j %>%
  filter(year >= 2017) %>%
  group_by(journal_group, usable) %>%
  summarize(n = n()) %>%
  ungroup() %>%
  pivot_wider(names_from = usable, values_from = n) %>%
  replace_na(list(no = 0, yes = 0)) %>%
  mutate(fraction = yes / (yes + no), total = yes + no) %>%
  mutate(fraction = round(fraction, digits = 3)) %>%
  arrange(desc(fraction))

g <- ggplot(
  gds_j %>% filter(year >= 2017) %>% mutate(journal_group = factor(journal_group, levels = gds_j_frac$journal_group)),
  aes(x = journal_group, fill = usable)
) +
  geom_bar() +
  scale_fill_brewer() +
  geom_text(data = gds_j_frac, color = "black", aes(x = journal_group, y = total + 30, label = fraction, fill = NA), size = 3) +
  cowplot::theme_cowplot() +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 14),
    axis.text.y = element_text(size = 14),
    axis.title.x = element_text(size = 14),
    axis.title.y = element_text(size = 14)
  ) 
g
```

```{r journaldot}
gds_sum <- gds_j %>%
  filter(year >= 2017) %>%
  group_by(journal, journal_group, journal_group2) %>%
  summarize(n = n(), frac = sum(usable == "yes") / n) %>%
  ungroup()

g <- ggplot(
  gds_sum %>% filter(n >= 3) %>%
    mutate(journal_group = factor(journal_group,
      levels = c("CNS", "CNS sub", "BMC", "CSHL", "eLife", "PLOS", "other")
    )),
  aes(x = n, y = frac, color = journal_group)
) +
  geom_point() +
  scale_color_brewer(palette = "Accent") +
  cowplot::theme_cowplot() +
  xlab("total studies (since 2017)") +
  ylab("fraction with metadata") +
  ylim(c(0, 1)) +
  theme(
    axis.text.x = element_text(size = 14),
    axis.text.y = element_text(size = 14),
    axis.title.x = element_text(size = 14),
    axis.title.y = element_text(size = 14)
  ) +
  scale_x_log10()
g # fig2b

gds_b <- gds_j %>%
  filter(year >= 2017) %>%
  select(id, journal, journal_group)

# write_tsv(gds_b, "fig2b.tsv.gz")
```

# Potential effects of GEO cell-level metadata on data reuse and citations

```{r if}
# ifs <- scholar::get_impactfactor(sapply(gds_j$pubmed, function(x) str_remove(str_remove(x$journal[1], "^The |^the |^THE "), " \\(.+")))
# ifs_abbr <- scholar::get_impactfactor(gds_j$journal)
# cut <- ifs_abbr %>% filter(Journal %in% c("CELL", "NATURE", "SCIENCE")) %>% pull(ImpactFactor) %>% max()
# ifs2 <- bind_cols(if_long = ifs$ImpactFactor, if_short = ifs_abbr$ImpactFactor) %>%
#   mutate(if_long = ifelse(if_long > cut, NA, if_long),
#          if_short = ifelse(if_short > cut, NA, if_short)) %>%
#   mutate(ifs = pmax(if_long, if_short, na.rm = TRUE))
# gds_j$ifs <- ifs2$ifs
#
# checked_j <- gds_j %>% group_by(journal, ifs) %>% summarise(n = n()) %>%
#   mutate(ifs = case_when(
#     journal == "JCI Insight" ~ 6.205,
#     journal == "Science Immunology" ~ 8.160,
#     journal == "J Am Soc Nephrol" ~ 9.274,
#     journal == "Genes Dev" ~ 9.527,
#     journal == "Blood Adv" ~ 4.584,
#     journal == "Curr Biol" ~ 5.720,
#     journal == "Protein Cell" ~ 10.164,
#     journal == "iScience" ~ 4.447,
#     journal == "Nat Struct Mol Biol" ~ 11.98,
#     TRUE ~ ifs
#   ))
#
# write_csv(checked_j, here("inst", "extdata", "ifs.csv"))
checked_j <- read_csv(here("inst", "extdata", "ifs.csv")) %>% select(-n)
gds_j <- gds_j %>% left_join(checked_j)
```

```{r citations}
yearcut <- max(gds_j$year) - 2
addhalf <- gds_j %>%
  filter(year == yearcut + 2) %>%
  pull(month) %>%
  unique() %>%
  length()


gds_c_if2 <- gds_j %>%
  mutate(bin = case_when(
    ifs < 5 ~ "5",
    ifs < 10 ~ "10",
    ifs < 30 ~ "30",
    TRUE ~ "+",
  )) %>%
  filter(year >= 2016 & year <= yearcut)

gds_c_if2 <- gds_c_if2 %>% 
  select(pubmed, usable, journal, cite, bin, year, ifs) %>%
  distinct() %>%
  filter(!is.na(cite)) %>%
  mutate(cite = as.numeric(cite)) %>%
  mutate(bin = factor(bin, levels = c("5", "10", "30", "+")))

g <- ggplot(
  gds_c_if2,
  aes(x = ifs, y = cite, group = usable, color = usable, shape = usable)
) +
  facet_grid(. ~ year) +
  geom_point(alpha = 0.87) +
  geom_point(
    alpha = 0.87,
    data = gds_c_if2 %>% filter(usable == "yes")
  ) +
  cowplot::theme_cowplot() +
  geom_smooth(method = "lm", fill = NA) +
  scale_color_brewer() +
  coord_cartesian(ylim = c(0, 500)) +
  xlab("journal IF") +
  ylab("# of citations") 
g # fig2c

gds_c <- gds_j %>% filter(year >= 2016 & year <= 2018) %>% 
  select(id, usable, journal, cite, year, ifs) %>%
  distinct() %>%
  filter(!is.na(cite)) %>%
  mutate(cite = as.numeric(cite))

# write_tsv(gds_c, "fig3c.tsv.gz")
```

```{r}
library(rbiorxiv)
biorxiv_conv <- function(doi1) {
  tryCatch(biorxiv_content(doi = doi1)[[1]]$published,
    error = function(e) doi1
  )
}

classes <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/categories-idx.tsv")
classes <- split(classes$Tool, classes$Category)

d <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/doi-idx.tsv")
ref <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/references.tsv")

classes <- sapply(classes, function(x) {
  d %>%
    filter(Tool %in% x) %>%
    pull(DOI)
})
classes$all <- unlist(classes) %>% unique()
ts <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/references.tsv") %>%
  filter(!is.na(Title)) %>%
  mutate(DOI_pub = mapply(biorxiv_conv, DOI))
try2l <- list()

id_conv <- function(doi) {
  tryCatch(rcrossref::id_converter(doi)$records$pmid,
    error = function(e) NA
  )
}

extract_layer <- function(x) {
  temp <- x %>% unlist()
  if (is.null(temp)) {
    temp <- NA
  }
  temp
}
try2l <- ts %>%
  filter(DOI_pub != "NA") %>%
  select(DOI_pub) %>%
  mutate(pmid = mapply(id_conv, DOI_pub))
try2l <- try2l %>% 
  mutate(pmid = map_chr(pmid, extract_layer))

tools <- read_tsv("https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/references.tsv") %>%
  pull(Title) %>%
  str_remove_all(":|,|\\(|\\)") %>%
  str_replace_all("’s | or | and |–|—a |—", " ") %>%
  str_remove_all(" Fig\\. 1\\.") #%>% na.omit()

tau <- list()
for (i in 1:length(tools)) {
  temp <- tryCatch(easyPubMed::get_pubmed_ids_by_fulltitle(tools[i],
    api_key = "9cc836a53f5c509a6783283e694da4fb0d08"
  )$IdList$Id,
  error = function(e) {
    NA
  })
  if (!is.null(temp)) {
    tau[[i]] <- temp
  } else {
    tau[[i]] <- NA
  }
}

tau2 <- tau
tau2[sapply(tau2, is.null)] <- NA
ts$tau <- unlist(tau2)
ts2 <- ts %>%
  left_join(try2l, by = "DOI_pub") %>%
  mutate(pubmed = ifelse(is.na(pmid), tau, pmid)) %>%
  filter(!is.null(unlist(pubmed)), !is.na(unlist(pubmed))) %>% 
  select(DOI, pubmed) %>%
  na.omit() %>% 
  distinct() %>%
  column_to_rownames("DOI")
ts2 <- ts2[ts$DOI, ]
names(ts2) <- ts$DOI

pubmeds_t2 <- list()
ts2[sapply(ts2, is.null)] <- NA
ts2 <- ts2 %>% unlist()
splits <- split(
  ts2 %>% na.omit() %>% unique(),
  ceiling(seq_along(ts2 %>% na.omit() %>% unique()) / 20)
)
for (i in 1:length(splits)) {
  s <- splits[[i]]
  query_string <- str_c(s, collapse = ",")
  pubmed_file <- easyPubMed::get_pubmed_ids(query_string,
    api_key = "9cc836a53f5c509a6783283e694da4fb0d08"
  )
  pubmed <- pubmed_file %>%
    easyPubMed::fetch_pubmed_data(retmax = 9999) %>%
    easyPubMed::articles_to_list()
  pubmeds1 <- sapply(pubmed, easyPubMed::article_to_df, simplify = FALSE)
  names(pubmeds1) <- pubmed_file$IdList
  pubmeds_t2 <- append(pubmeds_t2, pubmeds1)
}

notold <- sapply(pubmeds_t2, function(x) {
  as.numeric(x$year[1])
}, simplify = TRUE)
pubmeds_t3 <- pubmeds_t2[notold >= 2017]

get_authors <- function(pubmed) {
  tryCatch(pubmed %>% mutate(fullname = str_c(firstname, lastname, sep = "_")) %>%
    pull(fullname),
  error = function(e) {
    "NA"
  }
  )
}
```

```{r}
gds_a <- gds %>% select(id, journal, usable, year, pubmed)
authors <- list()
for (i in 1:nrow(gds_a)) {
  authors[[i]] <- get_authors(gds$pubmed[i][[1]])
}
gds_a$authors <- authors

plot_author <- function(classy) {
  fir <- list()
  for (i in 1:length(pubmeds_t3)) {
    fir[[i]] <- get_authors(pubmeds_t3[[i]])
  }
  firs <- fir %>%
    unlist() %>%
    unique()
  firs <- setdiff(firs, c("NA"))

  pubmeds_tc <- pubmeds_t3[ts2[classy] %>% na.omit()]
  firc <- list()
  for (i in 1:length(pubmeds_tc)) {
    firc[[i]] <- get_authors(pubmeds_tc[[i]])
  }
  firsc <- firc %>%
    unlist() %>%
    unique()
  firsc <- setdiff(firsc, c("NA"))
  firsc_check <<- firsc

  fir_id <- gds_a %>%
    select(-pubmed) %>%
    unnest("authors") %>%
    group_by(id) %>%
    filter(authors %in% firs) %>%
    pull(id) %>%
    unique()

  fir_idc <- gds_a %>%
    select(-pubmed) %>%
    unnest("authors") %>%
    group_by(id) %>%
    filter(authors %in% firsc) %>%
    pull(id) %>%
    unique()

  fir_idc_check <<- fir_idc

  gds_at <- gds %>%
    filter(year >= 2017) %>%
    mutate(tool = case_when(
      id %in% fir_idc ~ "yes",
      TRUE ~ "no"
    )) %>%
    group_by(tool, usable2) %>%
    summarize(n = n()) %>%
    group_by(tool) %>%
    mutate(fraction = n / sum(n))

  frac3 <- gds_at %>%
    filter(usable2 == "no") %>%
    mutate(fraction = 1 - fraction) %>%
    mutate(fraction = round(fraction, digits = 3))

  g <- ggplot(
    gds_at %>% mutate(usable2 = factor(usable2, levels = c("no", "meta", "object"))) %>%
      mutate(tool = factor(tool, levels = c("no", "yes"))),
    aes(x = tool, fill = usable2, y = fraction)
  ) +
    geom_col() +
    scale_fill_brewer(name = "usable") +
    geom_text(data = frac3, color = "black", aes(y = 1.03, label = fraction), size = 3) +
    cowplot::theme_cowplot() +
    ylab("fraction") +
    theme(
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 14),
      axis.text.y = element_text(size = 14),
      axis.title.x = element_text(size = 14),
      axis.title.y = element_text(size = 14)
    )
  g
}
ps <- plot_author(classes$all)
```

# Investigation of whether metadata deposition is better in projects with authors developing scRNA-seq informatic tools (split out to "upload author/other author/not", or just "author/not")

```{r}
gds$upload <- sapply(gds$geo, function(x) {
  temp <- x$contact_name
  if (is.null(temp)) {
    return("NA")
  } else {
    temp
  }
})

gds_cor <- gds_a %>%
  unnest(authors) %>%
  group_by(id) %>%
  slice(n())

gds_up <- gds %>%
  select(id, usable2, upload, year) %>%
  separate(upload, ",", into = c("F", "M", "L")) %>%
  mutate(full1 = str_c(str_to_title(F), " ", str_to_upper(M), "_", str_to_title(L))) %>%
  mutate(M = str_sub(M, 1, 1)) %>%
  mutate(full2 = str_c(str_to_title(F), " ", str_to_upper(M), "_", str_to_title(L))) %>%
  mutate(full3 = str_c(str_to_title(F), "_", str_to_title(L))) %>%
  mutate(check = case_when(
    (full1 %in% firsc_check) | (full2 %in% firsc_check) | (full3 %in% firsc_check) ~ "upload",
    id %in% fir_idc_check ~ "author",
    TRUE ~ "no"
  )) %>%
  left_join(gds_cor %>% select(id, authors))

frac4 <- gds_up %>%
  na.omit() %>%
  mutate(check = case_when(
    (full1 %in% firsc_check) | (full2 %in% firsc_check) | (full3 %in% firsc_check) ~ "upload",
    id %in% fir_idc_check ~ "author",
    TRUE ~ "no"
  )) %>% 
  group_by(year, usable2, check) %>%
  summarize(n = n()) %>%
  ungroup() %>%
  pivot_wider(names_from = usable2, values_from = n) %>%
  replace_na(list(no = 0, meta = 0, object = 0)) %>%
  mutate(fraction = (meta + object) / (meta + object + no), total = meta + object + no) %>%
  mutate(fraction = round(fraction, digits = 3)) %>%
  mutate(check = factor(check, levels = c("upload", "author", "no")))

g <- ggplot(frac4 %>% filter(year >= 2017), aes(x = year, y = fraction, color = check)) +
  geom_line() +
  geom_point() +
  scale_color_brewer(palette = "Accent") +
  cowplot::theme_cowplot() +
  xlab("year") +
  ylab("fraction with metadata") +
  ylim(c(0, 1)) +
  theme(
    axis.text.x = element_text(size = 14),
    axis.text.y = element_text(size = 14),
    axis.title.x = element_text(size = 14),
    axis.title.y = element_text(size = 14)
  )

g
```

```{r eval = F}
authors_df <- gds_a %>%
  select(-pubmed) %>%
  filter(year >= 2017) %>%
  unnest("authors") %>%
  group_by(authors) %>%
  summarize(n = n(), yes = sum(usable != "no")) %>%
  mutate(frac = yes / n)

# write_tsv(authors_df, "authors_meta.tsv")
```

```{r}
frac4 <- gds %>%
  na.omit() %>%
  mutate(check = case_when(
    id %in% fir_idc_check ~ "author",
    TRUE ~ "none"
  )) %>% 
  filter(year >= 2017) %>% 
  group_by(usable2, check) %>%
  summarize(n = n()) %>%
  group_by(check) %>% 
  mutate(fraction = n/sum(n))
  
frac5 <- frac4 %>%  
  ungroup() %>%
  select(-fraction) %>% 
  pivot_wider(names_from = usable2, values_from = n) %>%
  replace_na(list(no = 0, meta = 0, object = 0)) %>%
  mutate(fraction = (meta + object) / (meta + object + no), total = meta + object + no) %>%
  mutate(fraction = round(fraction, digits = 3))

frac6 <- frac4 %>% filter(usable2 == "no") %>% 
  mutate(fraction = 1 - fraction) %>% 
  mutate(fraction = round(fraction, digits = 3))

g <- ggplot(frac4 %>% mutate(usable2 = factor(usable2, levels = c("no", "meta", "object"))) %>% 
              mutate(check = factor(check, levels = c("none", "author"))),
       aes(x = check, fill = usable2, y = fraction)) +
  geom_col() +
  scale_fill_brewer(name = "usable") +
  geom_text(data = frac6, color = "black", aes(y = 1.03, label = fraction), size = 3) +
  cowplot::theme_cowplot() +
  ylab("fraction") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 14), 
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14))
# cowplot::save_plot("geo_authors.pdf", g, base_asp = 1, base_width = 5)
g # fig2d

gds_d <- gds %>%
  na.omit() %>%
  mutate(check = case_when(
    id %in% fir_idc_check ~ "author",
    TRUE ~ "none"
  )) %>% 
  dplyr::filter(year >= 2017) %>% 
  dplyr::select(id, usable, usable2, software_author = check)

# write_tsv(gds_d, "fig2d.tsv.gz")
```

<!-- # Similar analysis and observations with ArrayExpress -->

```{r ae1, eval = FALSE}
# if (format(format = "%m%d%y", file.info(here("inst", "extdata", "current_geo.rds"))$ctime) == date) {
if (!(file.exists(here("inst", "extdata", date, paste0("gds_result_", date, ".txt"))))) {
  sets <- read_tsv(here("inst", "extdata", "arrayexpress_sets_proc_add.tsv.gz"))
} else {
  source(here("utils", "arrayexpress_query.R"))
  sets <- read_tsv(here("inst", "extdata", "arrayexpress_sets_proc_add.tsv.gz"))
}

# remove GEO imported
sets2 <- sets %>% filter(!str_detect(ID, "GEO"))

sets_10x <- sets2 %>% filter(str_to_lower(str_sub(samplemetainfo_2, 1, 3)) == "10x")

sets_10x_withmeta <- sets_10x %>% 
  filter(
    str_detect(str_to_lower(samplemetainfo_1), "meta|annot|type|clustering|coldata") | 
    str_detect(str_to_lower(additionalfiles), "meta|annot|type|clustering|coldata")
  )

print("experiments")
sets2$samplemetainfo_2 %>% table() %>%
  as.data.frame() %>%
  arrange(desc(Freq))
```

```{r ae2, eval = FALSE}
print("fraction with metadata (10x experiments only)")
sets_10x %>% separate(ReleaseDate, sep = "-", into = c("year", "month", "day")) %>% 
  group_by(year) %>% 
  summarize(frac = sum(
    str_detect(str_to_lower(samplemetainfo_1), "meta|annot|type|clustering|coldata") | 
    str_detect(str_to_lower(additionalfiles), "meta|annot|type|clustering|coldata"),
    na.rm = TRUE) / n(),
    n = n())
```
