---
title: 'single cell GEO records'
date: '`r Sys.Date()`'
# package: scmetadata
author:
  - name: Rui Fu
    affiliation: RNA Bioscience Initative, University of Colorado School of Medicine
output:
  BiocStyle::html_document:
    toc_float: true

vignette: >
    %\VignetteIndexEntry{Introduction to clustifyr}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}  
---

```{r "knitr options", echo = FALSE, message = FALSE, warning = FALSE}
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  collapse = TRUE,
  fig.align = "center",
  comment = "#>",
  crop = NULL
)
```

```{r load}
library(here)
library(tidyverse)
date <- format(format = "%m%d%y", Sys.Date())
```

## Current GEO query used: 

`"expression profiling by high throughput sequencing" AND ("single nuclei" OR "single cell" OR "single-cell" OR "scRNAseq" OR "scRNA-seq" OR "snRNAseq" OR "snRNA-seq")`

## Parse GEO query

```{r entries}
#source(here("utils", "geo_query.R"))
gds <- read_tsv(here("inst", "extdata", paste0("geo_", "091020", ".tsv")))
gds_hm <- gds %>% filter(str_detect(org, "Homo sapiens") | str_detect(org, "Mus musculus"))
```
The fraction of GEO entries with potential metadata (file with "meta" or "annot" in filename or rda/rds/rdata files) is `r (gds_hm$usable == "yes") %>% mean()`

## Parse database from Svensson et al.

```{r curated}
library(googlesheets4)
gs4_deauth()
data <- read_sheet("https://docs.google.com/spreadsheets/d/1En7-UV0k0laDiIfjFkdn7dggyR7jIk3WH8QgXaMOZF0", sheet = "Data")
hm_data <- data %>% filter(Measurement == "RNA-seq") %>% filter(str_detect(`Data location`, "^GSE")) %>% filter(str_detect(Organism, "Human") | str_detect(Organism, "Mouse"))
hm_ids <- hm_data$`Data location` %>% strsplit(", ") %>% unlist()
```

```{r matchsub}
library(GEOquery)
get_subseries <- function(geo) {
  res <- tryCatch(
    suppressMessages(getGEO(GEO = geo,
                            filename = NULL, destdir = tempdir(), 
                            GSElimits = NULL, GSEMatrix = FALSE, 
                            AnnotGPL = FALSE, getGPL = TRUE,
                            parseCharacteristics = TRUE)),
    error = function(e) {"notavail"})
  if (class(res) != "GSE") {
    return(NA)
  }
  resr <- res@header$relation
  resr[str_detect(resr, "SuperSeries of: ")] %>% str_remove("SuperSeries of: ")
}

checksub <- purrr::map(
  setdiff(hm_ids, gds_hm$id),
  function(x) {
    temp <- get_subseries(x)
    if (length(temp) == 1) {
      if (is.na(temp)) {
        NA
      }
      else {
        length(intersect(temp, gds_hm$id)) == 0
      }
    } else {
      length(intersect(temp, gds_hm$id)) == 0
    }
  }) %>% unlist()
```
Overlap of human and mouse scRNA-seq GEO entries between manual curation and the GEO query is `r 
1 - sum(checksub, na.rm = T)/(length(hm_ids) - sum(is.na(checksub)))`.

Number of entries not public: `r sum(is.na(checksub))`

```{r date, include = FALSE, eval = F}
get_date <- function(geo) {
  res <- tryCatch(
    suppressMessages(GEOquery::getGEO(GEO = geo,
                            filename = NULL, destdir = tempdir(), 
                            GSElimits = NULL, GSEMatrix = FALSE, 
                            AnnotGPL = FALSE, getGPL = FALSE,
                            parseCharacteristics = FALSE)),
    error = function(e) {"notavail"})
  if (class(res) != "GSE") {
    return(NA)
  }

  res@header$submission_date
}

#library(progress)
# pb <- progress_estimated(nrow(gds_hm[1:1000,]))
# gds_hm2a <- gds_hm[1:1000,] %>%
#   mutate(date = map(id, function(x) {
#     # update the progress bar (tick()) and print progress (print())
#     pb$tick()$print()
#     get_date(x)
#   })) %>%
#   separate(date, into = c("month","day","year"))
# 
# pb <- progress_estimated(nrow(gds_hm[1001:2000,]))
# gds_hm2b<- gds_hm[1001:2000,] %>%
#   mutate(date = map(id, function(x) {
#     # update the progress bar (tick()) and print progress (print())
#     pb$tick()$print()
#     get_date(x)
#   })) %>%
#   separate(date, into = c("month","day","year"))
# pb <- progress_estimated(nrow(gds_hm[2001:3705,]))
# dates <- sapply(gds_hm$id[2001:3705], function(x) {
#   pb$tick()$print()
#   get_date(x)
# }, simplify = FALSE)
# saveRDS(dates, "3705dates.rds")
# gds_hm2c <- gds_hm[2001:3705,] 
# gds_hm2c$date <- unlist(dates)
# gds_hm2a <- readRDS(here("hm2000.rds"))
# gds_hm2 <- bind_rows(gds_hm2a, 
#                      gds_hm2c %>%
#   separate(date, into = c("month","day","year")))
```

```{r frac}
frac <- gds_hm %>% na.omit() %>% 
  group_by(year, usable) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  pivot_wider(names_from = usable, values_from = n) %>% 
  replace_na(list(no = 0, yes = 0)) %>% 
  mutate(fraction = yes/(yes + no), total = yes + no) %>% 
  mutate(fraction = round(fraction, digits = 3))

ggplot(gds_hm %>% na.omit(),
       aes(x = year, fill = usable)) +
  geom_bar() +
  scale_fill_viridis_d(option = "plasma") +
  geom_text(data = frac, color = "black", aes(x = year, y = total + 30, label = fraction, fill = NA)) +
  cowplot::theme_cowplot() +
  ggtitle("scRNAseq datasets by year")

ggplot(gds_hm %>% na.omit() %>% filter(year == "2020") %>% 
         mutate(month = factor(month, levels = c("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"))),
       aes(x = month, fill = usable)) +
  geom_bar() +
  scale_fill_viridis_d(option = "plasma") +
  cowplot::theme_cowplot() +
  ggtitle("scRNAseq datasets by month in 2020")
```

```{r cj, include = FALSE, eval = F}
ids_nat <- hm_data %>% filter(Journal == "Nature" | str_detect(Journal, "^Nat "))
ids_cell <- hm_data %>% filter(Journal %in% c("Cell","Cell Reports", "Developmental Cell", "Cell Stem Cell", "Molecular Cell", "Cancer Cell", "Immunity", "Neuron"))
ids_sci <- hm_data %>% filter(Journal %in% c("Science", "Sci Rep"))
ids_elife <- hm_data %>% filter(Journal == "eLife")
ids_pnas <- hm_data %>% filter(Journal == "Proc Natl Acad Sci USA")
ids_gb <- hm_data %>% filter(Journal == "Genome Biol")

gds_hm_j <- gds_hm %>%
  mutate(journal = case_when(
    id %in% ids_nat$`Data location` ~ "Nat",
    id %in% ids_cell$`Data location` ~ "Cell",
    id %in% ids_sci$`Data location` ~ "Sci",
    id %in% ids_elife$`Data location` ~ "eLife",
    id %in% ids_pnas$`Data location` ~ "PNAS",
    id %in% ids_gb$`Data location` ~ "GB",
    id %in% hm_data$`Data location` ~ "other"
    )) %>% 
  na.omit()

gds_hm_j_frac <- gds_hm_j %>% 
  group_by(journal, usable) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  pivot_wider(names_from = usable, values_from = n) %>% 
  replace_na(list(no = 0, yes = 0)) %>% 
  mutate(fraction = yes/(yes + no), total = yes + no) %>% 
  mutate(fraction = round(fraction, digits = 3))

ggplot(gds_hm_j %>% na.omit(),
       aes(x = journal, fill = usable)) +
  geom_bar() +
  scale_fill_viridis_d(option = "plasma") +
  geom_text(data = gds_hm_j_frac, color = "black", aes(x = journal, y = total + 10, label = fraction, fill = NA)) +
  cowplot::theme_cowplot() +
  ggtitle("scRNAseq datasets by journal")
```

```{r getjournal, include = F, eval = F}
library(GEOquery)
library(easyPubMed)
get_journal <- function(id) {
  res <- tryCatch(
    suppressMessages(GEOquery::getGEO(GEO = id,
                            filename = NULL, destdir = tempdir(), 
                            GSElimits = c(0,0), GSEMatrix = FALSE, 
                            AnnotGPL = FALSE, getGPL = FALSE,
                            parseCharacteristics = FALSE)),
    error = function(e) {"notavail"})
  if (class(res) != "GSE") {
    return(NA)
  }

  if (length(res@header$pubmed_id) > 1) {
      id2 <- tryCatch(res@header$pubmed_id[1],
                      error = function(e) {"notfound"}
      )
  } else {
      id2 <- tryCatch(res@header$pubmed_id,
                      error = function(e) {"notfound"}
      )
  }
  
  if (is.null(id2)) {
    return(NA)
  }
  if (id2 == "notfound") {
    return(NA)
  }
  
  tryCatch(
    get_pubmed_ids(id2) %>% 
      fetch_pubmed_data() %>%
      article_to_df() %>%
      dplyr::slice(1) %>%
      pull(jabbrv),
    error = function(e) "error"
  )
}

get_journal("GSE121386")

library(progress)
pb <- progress_estimated(nrow(gds_hm[2001:3000,]))
journal <- sapply(gds_hm$id[2001:3000], function(x) {
  pb$tick()$print()
  get_journal(x)
}, simplify = FALSE)

saveRDS(journal, "3705journals_3b.rds")

# journal <- list()
pb <- progress_estimated(nrow(gds_hm[3632:3705,]))
for (i in 632:705) {
  pb$tick()$print()
  journal[i] <- get_journal(gds_hm$id[3000 + i])
}
saveRDS(journal, "3705journals_4.rds")

gds_hm$journal <- c(readRDS("3705journals_1.rds") %>% unlist(),
readRDS("3705journals_2.rds") %>% unlist(),
readRDS("3705journals_3b.rds") %>% unlist(),
readRDS("3705journals_4.rds") %>% unlist())
```

```{r journalplot}
gds_hm_j <- gds_hm %>%
  mutate(journal_group = case_when(
    journal %in% c("Nature", "Science", "Cell") ~ "CNS",
    journal %in% c("Nat Commun", "Nat Metab", "iScience", "Cell Res.", "Sci Rep", "Sci Immunol", "Sci Adv", "Cell Rep", "Dev. Cell", "Cell Stem Cell", "Stem Cell Reports", "Mol. Cell", "Cancer Cell", "Immunity", "Neuron", "Cell Syst", "Cell Metab.", "Cell Host Microbe") ~ "CNS sub",
    str_detect(journal, "^Nat. ") ~ "CNS sub",
    journal %in% c("Genome Biol.", "Genome Med") ~ "BMC",
    str_detect(journal, "BMC") ~ "BMC",
    journal %in% c("Genes Dev.", "Genome Res.") ~ "CSHL",
    journal %in% c("Elife") ~ "eLife",
    journal %in% c("Proc. Natl. Acad. Sci. U.S.A.") ~ "PNAS",
    str_detect(journal, "PLoS") ~ "PLOS",
    #journal %in% c("Nucleic Acids Res.") ~ "NAR",
    
    TRUE ~ "other"
    )) %>% 
  filter(journal != "error") %>% 
  filter(year >= 2018) %>% 
  na.omit()

gds_hm_j_frac <- gds_hm_j %>% 
  group_by(journal_group, usable) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  pivot_wider(names_from = usable, values_from = n) %>% 
  replace_na(list(no = 0, yes = 0)) %>% 
  mutate(fraction = yes/(yes + no), total = yes + no) %>% 
  mutate(fraction = round(fraction, digits = 3)) %>% 
  arrange(desc(fraction))

ggplot(gds_hm_j %>% mutate(journal_group = factor(journal_group, levels = gds_hm_j_frac$journal_group)),
       aes(x = journal_group, fill = usable)) +
  geom_bar() +
  scale_fill_viridis_d(option = "plasma") +
  geom_text(data = gds_hm_j_frac, color = "black", aes(x = journal_group, y = total + 30, label = fraction, fill = NA)) +
  cowplot::theme_cowplot() +
  ggtitle("scRNAseq datasets by journal")
```

